Refactor and add tests and metric to KillUnusedSegments duty (auto-kill)

abhishekrb19 opened the pull request:
This PR makes a few functional and non-functional changes in the auto-kill `KillUnusedSegments` duty.

Functional changes:

- Add a new metric `kill/candidateUnusedSegments/count` with dimension `dataSource` that tracks the number of candidate unused segments per datasource per kill cycle. It's candidate because it's the kill task that ultimately kills the unused segments and the number can change.
- Move a debug log to info for improved visibility. This should be fine as it gets logged only once per indexing/kill task period.
- Change the kill duty DruidException validations from user to operator persona as they're configured by operators typically.

Refactoring changes:
- Remove redundant checks in the code for code readability.
- Remove all the unnecessary `@VisibleForTesting`  annotations as the test code now verifies the duty directly instead of mocks + various hooks into the code.
- Finalize some parameters and variables to clarify the intent.


Testing changes:
- Updated tests from using mocks (the mocks were difficult to use and buggy - for example, last updated timestamp was never used.
- Fix a bug where `TestDruidCoordinatorConfig.DEFAULT_COORDINATOR_KILL_BUFFER_PERIOD` was incorrectly set to 1 day instead of 90 days. While at it, I changed all the test coordinator defaults to match the same ISO 8601 period format as `DruidCoordinatorConfig`, so it's easy to spot check.
- Add more test coverage for the different config parameters.
- Add a couple of unit tests that are ignored currently as they fail. The duty doesn't consider segments spanning eternity, or end in `DateTimes.MAX` (outside the normal Datetime string comparison range): https://github.com/apache/druid/issues/15951


### Motivation
In general, I found it difficult to track the progress of the auto-kill coordinator duty. Also, see some discussion in the community slack: https://apachedruidworkspace.slack.com/archives/C0303FDCZEZ/p1707941058448649.
My preliminary conclusion from looking at a couple of clusters is that the auto-kill defaults are overly conservative and need to be revisited as https://github.com/apache/druid/issues/15912 notes.  I would like to address some of the concerns raised in https://github.com/apache/druid/issues/15911 and https://github.com/apache/druid/issues/15912 separately once we have some additional data and tests from this patch. 

#### Release note
<!-- Give your best effort to summarize your changes in a couple of sentences aimed toward Druid users. 

If your change doesn't have end user impact, you can skip this section.

For tips about how to write a good release note, see [Release notes](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#release-notes).

-->

Add a new metric `kill/candidateUnusedSegments/count` with dimension `dataSource` that tracks the number of candidate unused segments per datasource that are eligible for deletion during each auto kill cycle.


<hr>

##### Key changed/added classes in this PR
 * `KillUnusedSegments.java`
 * `KillUnusedSegmentsTest.java`

<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

kfaraz commented on server/src/test/java/org/apache/druid/server/coordinator/duty/KillUnusedSegmentsTest.java:
```diff
@@ -96,362 +104,805 @@ public class KillUnusedSegmentsTest
+    final DruidCoordinatorRuntimeParams runParams = killDuty.run(paramsBuilder.build());
+
+    validateStats(
+        new ExpectedStats(0, 0, 3, ImmutableMap.of()),
```
I feel it would be much more readable to just validate the relevant stats individually, rather than having to pass zeroes and empty maps repeatedly. This would allow someone reading the test to instantly recognize exactly what is being verified.

You can see some examples in `RunRulesTest`.
Something like:
```
Assert.assertEquals(10L, stats.getSegmentStat(Stats.Segments.ASSIGNED, "normal", DATASOURCE));
```

abhishekrb19 replied:
Yes, definitely, it reads more clearly now!

kfaraz reviewed (commented):
Overall, the changes make the code more readable. I couldn't find the main NPE bugfix amidst all the refactor. If possible, maybe create two separate PRs - one to add the metric and the bugfix and another for all the refactors?

I have left some comments even though the PR is still in draft, so that changes can be incorporated early.

kfaraz commented on server/src/test/java/org/apache/druid/server/coordinator/duty/KillUnusedSegmentsTest.java:
```diff
@@ -96,362 +104,805 @@ public class KillUnusedSegmentsTest
-        .when(overlordClient)
-        .taskStatuses(null, null, 0);
+
+    Assert.assertEquals(
+        expectedKillTaskId,
+        observedLastKillTaskId
+    );
```
Can be in a single line.

kfaraz commented on server/src/test/java/org/apache/druid/server/coordinator/duty/KillUnusedSegmentsTest.java:
```diff
@@ -96,362 +104,805 @@ public class KillUnusedSegmentsTest
+
+    createAndAddUnusedSegment(DS1, Intervals.ETERNITY, sixtyDaysAgo);
+
+    final KillUnusedSegments killDuty = new KillUnusedSegments(
+        segmentsMetadataManager,
+        overlordClient,
+        configBuilder.build()
+    );
```
Constructing a `KillUnusedSegments` can be commoned out into a method.

kfaraz commented on server/src/test/java/org/apache/druid/server/coordinator/duty/KillUnusedSegmentsTest.java:
```diff
@@ -96,362 +104,805 @@ public class KillUnusedSegmentsTest
+    }
+  }
+
+  private void validateAndResetState(final String dataSource, @Nullable final Interval expectedKillInterval)
```
Please update the method name to indicate what is being validated.

abhishekrb19 commented on server/src/test/java/org/apache/druid/server/coordinator/simulate/TestSegmentsMetadataManager.java:
```diff
@@ -34,33 +36,69 @@
+  /**
+   * Adds the used segment.
+   */
+  public void addSegment(DataSegment segment, boolean isUsed)
```
Removed it, I was trying to consolidate some of the methods, but decided to just keep them separate

abhishekrb19 commented on server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java:
```diff
@@ -282,41 +300,16 @@ private Interval findIntervalForKill(String dataSource, DateTime maxUsedStatusLa
 
-  @VisibleForTesting
-  static int getKillTaskCapacity(int totalWorkerCapacity, double killTaskSlotRatio, int maxKillTaskSlots)
+  private static int getKillTaskCapacity(int totalWorkerCapacity, double killTaskSlotRatio, int maxKillTaskSlots)
```
Yeah, good point. Removed `static` and also folded this function with the other helper.

abhishekrb19 commented on server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java:
```diff
@@ -146,130 +154,140 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
-          )
+              ImmutableList.copyOf(dataSourcesToKill).subList(submittedTasks, dataSourcesToKill.size()))
               : ""
-      );
```
Yeah, thanks, I wondered the same initially :) Updated per your suggestion

abhishekrb19 commented on server/src/test/java/org/apache/druid/server/coordinator/duty/KillUnusedSegmentsTest.java:
```diff
@@ -96,362 +104,805 @@ public class KillUnusedSegmentsTest
+
+    validateAndResetState(DS1, null);
+
+    log.info("Fourth run...");
```
Ack, removed the test logger altogether.

abhishekrb19 commented on server/src/test/java/org/apache/druid/server/coordinator/duty/KillUnusedSegmentsTest.java:
```diff
@@ -96,362 +104,805 @@ public class KillUnusedSegmentsTest
+  }
+
+  /**
+   * The kill period is honored after the first indexing run.
```
Updated the test name. Also, I included java docs only for tests that aren't very straightforward.

abhishekrb19 commented:
Thanks for the review, @kfaraz! I've addressed your comments and kept the new metric along with other minor functional changes in this patch. But let me know if you'd prefer to keep the changes separate.  Re the NPE fix, my bad, it was from existing test code --`TestSegmentsMetadataManager` was returning null instead of an empty set for `retrieveAllDataSourceNames()`. Removed that from the PR description.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java:
```diff
@@ -87,29 +90,36 @@ public class KillUnusedSegments implements CoordinatorDuty
-      );
+      throw DruidException.forPersona(DruidException.Persona.OPERATOR)
+                          .ofCategory(DruidException.Category.INVALID_INPUT)
+                          .build(StringUtils.format(
```
Nit: would look neater like this:
```suggestion
                          .build(
                              StringUtils.format(
```

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java:
```diff
@@ -146,130 +156,128 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
-        }
+    for (String dataSource : dataSourcesToKill) {
+      if (submittedTasks >= availableKillTaskSlots) {
+        log.info(StringUtils.format(
+            "Submitted [%d] kill tasks and reached kill task slot limit [%d]. Will resume "
+            + "on the next coordinator cycle.", submittedTasks, availableKillTaskSlots));
```
```suggestion
        log.info(
            "Submitted [%d] kill tasks and reached kill task slot limit [%d].",
            submittedTasks, availableKillTaskSlots
        );
```

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java:
```diff
@@ -146,130 +156,128 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
-    }
+    log.info(
+        "Submitted [%d] kill tasks for [%d] datasources. Remaining datasources to kill: %s",
+        submittedTasks,
+        dataSourcesToKill.size(),
+        remainingDatasourcesToKill
```
Super nit: I generally prefer this style as the logging occupies fewer lines without compromising readability, but either way is fine really.
```suggestion
        submittedTasks, dataSourcesToKill.size(), remainingDatasourcesToKill
```

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java:
```diff
@@ -146,130 +156,128 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
   /**
+   * <p>
    * Calculates the interval for which segments are to be killed in a datasource.
+   * Because this function uses string comparisons for date time, it doesn't find unused segments that are outside
```
```suggestion
   * Since this method compares datetime as strings, it cannot find unused segments that are outside
```

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java:
```diff
@@ -146,130 +156,128 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
    * Calculates the interval for which segments are to be killed in a datasource.
+   * Because this function uses string comparisons for date time, it doesn't find unused segments that are outside
+   * the range [{@link DateTimes#COMPARE_DATE_AS_STRING_MIN}, {@link DateTimes#COMPARE_DATE_AS_STRING_MAX}),
+   * such as {@link org.apache.druid.java.util.common.granularity.Granularities#ALL} partitioned segments and segments that
+   * end in {@link DateTimes#MAX}.
```
```suggestion
   * such as {@link org.apache.druid.java.util.common.granularity.Granularities#ALL} partitioned segments
   * and segments that end in {@link DateTimes#MAX}.
```

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java:
```diff
@@ -146,130 +156,128 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+        maxUsedStatusLastUpdatedTime
+    );
+
+    // Each unused segment interval returned here has a 1:1 correspondence with an unused segment. So we can assume
```
How do we know this? A single interval can have many killable unused segments.

abhishekrb19 replied:
Please see `getUnusedSegmentIntervals()` - it returns a list of unused segment intervals when it could actually just be the unused segments itself, but more lightweight computation. Also, added `testKillMultipleSegmentsInSameInterval()` to demonstrate the scenario.

I considered changing it to directly fetch the unused segments instead of the intervals, which would also allow us to compute any additional statistics like size, etc. If you think that's useful, I can do it in a follow-up patch.

kfaraz replied:
I see, thanks for the clarification! Looking at the implementation `SqlSegmentsMetadataManager`, that does indeed seem to be the case. I guess the javadoc in `SegmentsMetadataManager` didn't really call it out and I just assumed that it would be a list of unique intervals.

Although, it is a little weird for this API to return duplicate intervals, maybe something we should revisit? Or atleast call out in the `SegmentsMetadataManager` javadocs for now?

abhishekrb19 replied:
Yeah, I think we can revisit that later. Let me clarify the current behavior in the method's javadocs to avoid any confusion

kfaraz commented on docs/operations/metrics.md:
```diff
@@ -342,6 +342,7 @@ These metrics are for the Druid Coordinator and are reset each time the Coordina
 |`killTask/availableSlot/count`| Number of available task slots that can be used for auto kill tasks in the auto kill run. This is the max number of task slots minus any currently running auto kill tasks.                                                                                                                                                                                                    ...
 |`killTask/maxSlot/count`| Maximum number of task slots available for auto kill tasks in the auto kill run.                                                                                                                                                                                                                                                                                                     ...
 |`kill/task/count`| Number of tasks issued in the auto kill run.                                                                                                                                                                                                                                                                                                                                                ...
+|`kill/candidateUnusedSegments/count`|Number of candidate unused segments to be deleted from the metadata store in an auto kill run.|`dataSource`|Varies|
```
Reading at this doc and the name of the metric, it's not very clear what it represents.
Is it the number of segments that the coordinator _decided to kill_ in a datasource or the number of segments that _should be killed_ in a datasource? Also, should it have `interval` as a dimension?

abhishekrb19 replied:
> Is it the number of segments that the coordinator decided to kill in a datasource or the number of segments that should be killed in a datasource? 

It's the latter - clarified the docs a wee bit. Let me know if you have a better suggestion. I called it candidate because it's ultimately the underlying kill task that can say this for sure.

>Also, should it have interval as a dimension?

No, there can be multiple intervals.

kfaraz replied:
> No, there can be multiple intervals, so an interval dimension can explode the metric.

Makes sense.

kfaraz commented on server/src/test/java/org/apache/druid/server/coordinator/simulate/TestSegmentsMetadataManager.java:
```diff
@@ -221,7 +242,29 @@ public List<Interval> getUnusedSegmentIntervals(
+        }
+      }
+    }
+    log.info("Found [%d] unused segment intervals: [%s]", unusedSegmentIntervals.size(), unusedSegmentIntervals);
```
If this is really needed, maybe it should be a log line in `KillUnusedSegments` duty because `SqlSegmentsMetadataManager` doesn't seem to log this either.

abhishekrb19 replied:
Removed it as the new metric should suffice

kfaraz commented on server/src/test/java/org/apache/druid/server/coordinator/simulate/TestSegmentsMetadataManager.java:
```diff
@@ -124,8 +139,14 @@ public int markAsUnusedSegmentsInInterval(String dataSource, Interval interval)
   {
     int numModifiedSegments = 0;
+    final DateTime now = DateTimes.nowUtc();
+    final DateTime lastUpdatedDate = now.plus(10);
```
Please add a comment for why plus 10?

abhishekrb19 replied:
Removed the magic number :)

kfaraz reviewed (commented):
Thanks for the prompt response, @abhishekrb19 !
Left a few minor comments, rest of the changes look good.

kfaraz commented on docs/operations/metrics.md:
```diff
@@ -342,7 +342,7 @@ These metrics are for the Druid Coordinator and are reset each time the Coordina
 |`killTask/maxSlot/count`| Maximum number of task slots available for auto kill tasks in the auto kill run.                                                                                                                                                                                                                                                                                                     ...
 |`kill/task/count`| Number of tasks issued in the auto kill run.                                                                                                                                                                                                                                                                                                                                                ...
-|`kill/candidateUnusedSegments/count`|Number of candidate unused segments to be deleted from the metadata store in an auto kill run.|`dataSource`|Varies|
+|`kill/candidateUnusedSegments/count`|The number of candidate unused segments eligible for deletion from the metadata store during an auto kill run for a datasource.|`dataSource`|Varies|
```
```suggestion
|`kill/eligibleUnusedSegments/count`|The number of unused segments of a datasource that are identified as eligible for deletion from the metadata store by the coordinator.|`dataSource`|Varies|
```

kfaraz reviewed (approved):
Only 1 minor suggestion regarding docs and metric name. But that change can be done later and need not block this PR. It would trigger the whole CI pipeline all over again.

+1

abhishekrb19 commented:
Thanks, @kfaraz! I will add your rename suggestion in a follow-up patch.
